File: Antelope.cpp
#include "include/Antelope.h"
#include "include/Turtle.h"
#include "include/Plant.h"

Antelope::Antelope(World* world, Point& position)
: Animal(world, 4, 4, position, "🐐") {
}

Antelope::Antelope(World* world, Point& position, int age)
: Animal(world, 4, 4, position, "🐐", age) {
}

Antelope::~Antelope() {

}

void Antelope::action() {
    if (hasActed) return;
    Point destination = world->getRandomSpaceDoubleMove(position);
    if(world->isWithinBounds(destination) && age != 0) {
        Organism* other = world->getAtCoordinates(destination);

        if(haveSavedAttack(*other)) {
            return;
        }
        int status = -1;
        if(other != nullptr) {
            status = collision(*other);
        }
        if (status != 1) 
            move(destination);
    }

    age++;
    setHasActed(true);
}

bool Antelope::haveSavedAttack(Organism& other) {
    if (dynamic_cast<const Turtle*>(&other) != nullptr && this->getStrength() < 5) {
        return true;
    }
    return false;
}

int Antelope::escapeCollision(Organism& other) {
    int randomNumber = rand() % 2 + 1;
    if((randomNumber != 0) && (typeid(other) != typeid(*this)) && (dynamic_cast<Plant*>(&other) == nullptr)) {
        Point destination = world->getRandomFreeSpaceAround(position);
        if(world->isWithinBounds(destination)) {
            move(destination);
            setHasActed(true);
            return 1;
        }
    }
    return 0;
}

int Antelope::collision(Organism& other) {
    if(escapeCollision(other) == 1) {
        return 1;
    }
    if(canKill(other)) {
        kill(other);
        return 0;
    }
    if (typeid(other) == typeid(*this)) {
        if(canReproduce(other, position)) {
            reproduce(position);
            setHasActed(true);
        }
        return 1;
    }
    other.collision(*this);
    setHasActed(true);
    return 0;
}

File: game-of-life-java/src/Antelope.java
import javax.swing.*;
import java.awt.*;

public class Antelope extends Animal {
    private static final ImageIcon antelopeIcon = new ImageIcon("images/antelope.png");
    public static final Image scaledAntelopeIcon = antelopeIcon.getImage().getScaledInstance(30, 30, Image.SCALE_SMOOTH);

    public Antelope(Point position, World world, int age) {
        super(4, 4, position, world, age);
    }

    @Override
    public ImageIcon getIcon() {
        return antelopeIcon;
    }


    @Override
    protected int escapeCollision(Organism other) {
        int randomNumber = (int) (Math.random() * 2 + 1);
        if (randomNumber == 1 && !other.getClass().equals(this.getClass()) && !(other instanceof Plant)) {
            Point destination = world.getRandomFreeSpaceAround(position);
            if (world.isWithinBounds(destination)) {
                position = destination;
                setHasActed(true);
                return 1;
            }
        }
        return 0;
    }

    @Override
    protected boolean haveSavedAttack(Organism other) {
        if(other instanceof Turtle && this.getStrength() < 5) {
            return true;
        }
        return false;
    }

    @Override
    public void move() {
        if(getHasActed()) {
            return;
        }
        Point newPos = world.getRandomSpaceDoubleMove(position);
        if (newPos != null) {
            Organism other = world.getOrganismAtPosition(newPos);
            if(haveSavedAttack(other)) {
                return;
            }
            if (other != null) {
                collision(other);
                setHasActed(true);
            } else {
                position = newPos;
            }
        }
    }
}
